### PR TITLE
[dattri.func] rename ihvp.py -> hessian.py/fisher.py

### PR DESCRIPTION
**Pull request by TheaperDeng**
### 1. Motivation and Context
This is a renaming PR that will split the original ihvp collection to hessian and fisher separately to make them clearer in math perspective.

### 2. Summary of the change
1. Rename and split `dattri.func.ihvp` to `dattri.func.hessian` and `dattri.func.fisher`. Together with the unittest, related import command and document.
```python
# dattri.func.fisher
'''
This module contains:
- `ifvp_explicit`: IFVP via explicit FIM calculation.
- `ifvp_at_x_explicit`: IFVP via explicit FIM calculation (with fixed x).
- `ifvp_datainf`: DataInf ifvp algorithm function.
- `ifvp_at_x_datainf`: DataInf ifvp algorithm function (with fixed x).
- `ifvp_at_x_ekfac`: EK-FAC ifvp algorithm function (with fixed x).
'''
```
```python
# dattri.func.hessian
"""
This module contains:
- `hvp`: Calculate the Hessian Vector Product (HVP) of a function.
- `hvp_at_x`: Calculate the Hessian Vector Product (HVP) of a function with fixed x.
- `hvp`: Calculate the HVP of a function.
- `hvp_at_x`: Calculate the HVP of a function with fixed x.
- `ihvp_at_x_explicit`: IHVP via explicit Hessian calculation.
- `ihvp_cg`: Conjugate Gradient Descent ihvp algorithm function.
- `ihvp_at_x_cg`: Conjugate Gradient Descent ihvp algorithm function with fixed x.
- `ihvp_arnoldi`: Arnoldi Iteration ihvp algorithm function.
- `ihvp_at_x_arnoldi`: Arnoldi Iteration ihvp algorithm function with fixed x.
- `ihvp_lissa`: Lissa algorithm ihvp function.
- `ihvp_at_x_lissa`: Lissa algorithm ihvp function with fixed x.
"""
```
2. Add `ifvp_explicit` and `ifvp_at_x_explicit`
- Add special description for the `func` (split the empirical FIM and true FIM)
- check it can work on a trained NN with >0.99 correlation with explicit Hessian ihvp.

### 3. What tests have been added/updated for the change?
- [x] Unit test: Typically, this should be included if you implemented a new function/fixed a bug.

### CODE REVIEW

File: dattri/func/fisher.py [hunk 1]
```diff
@@ -0,0 +1,855 @@
...
+import warnings
+from functools import wraps
+
+import torch
+from torch.func import grad, vmap
+
+
+def ifvp_explicit(
+    func: Callable,
+    argnums: int,
```
Comment (tingwl0122) [inline dattri/func/fisher.py:34]: default to 0?

File: dattri/func/fisher.py [hunk 2]
```diff
@@ -0,0 +1,855 @@
...
+    calculate the IFVP for any given vector.
+
+    Args:
+        func (Callable): A function taking one or more arguments and returning
+            a single-element Tensor. The FIM will be calculated based on
+            this function. Notably, this function should be negative log-likelihood
+            (e.g., cross-entropy loss) for classification tasks. If you want to
+            calculate the emperial FIM, you should use the groundtruth label for
+            the loss. If you want to calculate the true FIM, you should use the
+            predicted label for the loss.
```
Comment (tingwl0122) [inline dattri/func/fisher.py:51]: possibly provide some example func?

File: dattri/func/fisher.py [hunk 3]
```diff
@@ -0,0 +1,855 @@
...
+            v.T,
+        ).T
+
+    return _ifvp_explicit_func
+
+
+def ifvp_at_x_explicit(
+    func: Callable,
+    *x,
+    argnums: Union[int, Tuple[int, ...]],
```
Comment (tingwl0122) [inline dattri/func/fisher.py:90]: default to 0?

File: README.md [hunk 4]
```diff
@@ -69,7 +69,7 @@ Hessian-vector product (HVP), inverse-Hessian-vector product
 (IHVP) are widely used in data attribution methods. `dattri` provides efficient implementation to these operators by `torch.func`. This example shows how to use the CG implementation of the IHVP implementation.
```
[the remark this answers is not in the document]
Comment (jiaqima) [reply README.md:69]: I feel it's ok to leave it as is here. The README is not supposed to go through everything anyway.

File: dattri/func/hessian.py [hunk 5]
```diff
@@ -1,13 +1,18 @@
...
+IHVP (inverse hessian-vector product)
+HVP (hessian-vector product)
 
 This module contains:
-- `hvp`: Calculate the Hessian Vector Product (HVP) of a function.
-- `hvp_at_x`: Calculate the Hessian Vector Product (HVP) of a function with fixed x.
+- `hvp`: Calculate the HVP of a function.
+- `hvp_at_x`: Calculate the HVP of a function with fixed x.
 - `ihvp_at_x_explicit`: IHVP via explicit Hessian calculation.
 - `ihvp_cg`: Conjugate Gradient Descent ihvp algorithm function.
```
Comment (jiaqima) [inline dattri/func/hessian.py:10]: Let's change all the "ihvp" in comments to "IHVP"

**Review by jiaqima — APPROVED** [review]
LGTM

**Pull request merged**